File: go-backend-interview-prep/03_functions_and_methods/03_methods_and_receivers/main.go
package main

import "fmt"

type Rect struct {
	Width, Height float64
}

// Value receiver: operates on a copy.
func (r Rect) Area() float64 {
	return r.Width * r.Height
}

// Value receiver: cannot mutate the original.
func (r Rect) Perimeter() float64 {
	return 2 * (r.Width + r.Height)
}

// Pointer receiver: operates on the original.
func (r *Rect) Scale(factor float64) {
	r.Width *= factor
	r.Height *= factor
}

// Pointer receiver: mutates the original.
func (r *Rect) SetWidth(w float64) {
	r.Width = w
}

// --- Named type to add methods to a built-in type ---
type StringSlice []string

func (ss StringSlice) Join(sep string) string {
	result := ""
	for i, s := range ss {
		if i > 0 {
			result += sep
		}
		result += s
	}
	return result
}

func main() {
	// --- Value receiver methods ---
	r := Rect{Width: 10, Height: 5}
	fmt.Println("area:", r.Area())
	fmt.Println("perimeter:", r.Perimeter())

	// --- Pointer receiver: mutates ---
	fmt.Println("\n--- Pointer receiver ---")
	fmt.Println("before scale:", r)
	r.Scale(2)
	fmt.Println("after Scale(2):", r) // {20, 10}

	r.SetWidth(100)
	fmt.Println("after SetWidth(100):", r)

	// --- Value receiver CANNOT mutate ---
	fmt.Println("\n--- Value receiver cannot mutate ---")
	r2 := Rect{Width: 3, Height: 4}
	tryToMutateValue(r2)
	fmt.Println("r2 unchanged:", r2)

	// --- Auto address-taking ---
	// Go lets you call a pointer-receiver method on a value.
	// The compiler automatically takes the address.
	fmt.Println("\n--- Auto address-taking ---")
	val := Rect{Width: 5, Height: 5}
	val.Scale(3) // compiler does (&val).Scale(3)
	fmt.Println("after val.Scale(3):", val)

	// --- Method on named type ---
	fmt.Println("\n--- Named type method ---")
	tags := StringSlice{"go", "python", "rust"}
	fmt.Println("joined:", tags.Join(", "))
}

func tryToMutateValue(r Rect) {
	r.Width = 999 // modifies the copy
}

File: go-backend-interview-prep/03_functions_and_methods/03_methods_and_receivers/main.py
# Python equivalent of Methods and Receivers (compare with main.go)
# Python methods always receive 'self' (a reference), similar to Go pointer receivers.
# There is no "value receiver" concept in Python.


class Rect:
    def __init__(self, width: float = 0, height: float = 0):
        self.width = width
        self.height = height

    def area(self) -> float:
        """All Python methods are like Go pointer receivers (self is a reference)."""
        return self.width * self.height

    def perimeter(self) -> float:
        return 2 * (self.width + self.height)

    def scale(self, factor: float) -> None:
        """Mutates the original (always, in Python)."""
        self.width *= factor
        self.height *= factor

    def set_width(self, w: float) -> None:
        self.width = w

    def __repr__(self):
        return f"Rect(width={self.width}, height={self.height})"


def try_to_mutate_value(r: Rect) -> None:
    """In Python, this WILL mutate the original (unlike Go value receiver)."""
    r.width = 999


def main():
    # --- Methods ---
    r = Rect(width=10, height=5)
    print("area:", r.area())
    print("perimeter:", r.perimeter())

    # --- Mutation (Python always mutates -- no value receiver concept) ---
    print("\n--- Mutation (always pointer-like) ---")
    print("before scale:", r)
    r.scale(2)
    print("after scale(2):", r)

    r.set_width(100)
    print("after set_width(100):", r)

    # --- Python functions ALWAYS get reference to mutable objects ---
    print("\n--- Function receives reference ---")
    r2 = Rect(width=3, height=4)
    try_to_mutate_value(r2)
    print("r2 CHANGED:", r2)  # width=999 (unlike Go value receiver)

    # Key difference:
    # Go value receiver = copy (no mutation)
    # Python self = reference (always mutates)
    print("\nKey: Go value receiver = copy. Python self = reference.")


if __name__ == "__main__":
    main()
